### PR TITLE
planner: set min for high risk plan steps

### PR DESCRIPTION
**Pull request by terry1purcell**
### What problem does this PR solve?

Issue Number: close #55126 

Problem Summary:

### What changed and how does it work?

Ensure a minimum number of rows and a minimum row size for each plan operation for the optimizer cost model. And ensure that non-index based operations have a minimum of 1 row processed, and row length of at least 2. Index based operations can be fractional (less than 1) due to index probes being averaged out across all outer to inner probes for joins - thus the minimum of 1 row does not apply to index operations due to risk of biasing away from index joins or index merges.

Estimates at zero or below zero are possible in many situations due to realtime insert/update/delete tracking for statistics processing is not transactional (loss of statistics are OK - since it is not user data, and will be corrected next ANALYE). Below zero is rare but possible. In cardinality estimation, zero or near zero it possible due to assumption of independence with multiple predicates or multiple joins, and also with underestimation of any (or many) individual predicate(s).

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/plan_cost_ver2.go [hunk 1]
```diff
@@ -103,6 +103,17 @@ func (p *PhysicalProjection) GetPlanCostVer2(taskType property.TaskType, option
 	return p.PlanCostVer2, nil
 }
 
+const (
+	// MinNumRows provides a minimum to avoid underestimation
+	MinNumRows = 1.0
+	// MinRowSize provides a minimum to avoid underestimation
+	MinRowSize = 2.0
```
Comment (qw4990) [inline pkg/planner/core/plan_cost_ver2.go:110]: Agree with your reason to set these 2 minimum for non-index based operations, but could you add code comments here to explain why we want to add them? For example, just moving some comments in the PR description to here as code comments.

Comment (terry1purcell) [reply pkg/planner/core/plan_cost_ver2.go]: Done

File: pkg/planner/core/casetest/partition/testdata/partition_pruner_out.json [hunk 2]
```diff
@@ -470,12 +470,12 @@
...
-          "  │ └─Selection 0.00 cop[tikv]  eq(test_partition.t2.b, 7), eq(test_partition.t2.id, 7), in(test_partition.t2.a, 6, 7, 8)",
-          "  │   └─TableFullScan 10000.00 cop[tikv] table:t2 keep order:false, stats:pseudo",
-          "  └─TableReader(Probe) 0.01 root partition:p0 data:Selection",
-          "    └─Selection 0.01 cop[tikv]  eq(test_partition.t1.id, 7), or(eq(test_partition.t1.a, 1), and(eq(test_partition.t1.a, 3), in(test_partition.t1.b, 3, 5)))",
-          "      └─TableFullScan 10000.00 cop[tikv] table:t1 keep order:false, stats:pseudo"
+          "  ├─TableReader(Build) 0.01 root partition:p0 data:Selection",
+          "  │ └─Selection 0.01 cop[tikv]  eq(test_partition.t1.id, 7), or(eq(test_partition.t1.a, 1), and(eq(test_partition.t1.a, 3), in(test_partition.t1.b, 3, 5)))",
+          "  │   └─TableFullScan 10000.00 cop[tikv] table:t1 keep order:false, stats:pseudo",
+          "  └─TableReader(Probe) 0.00 root partition:p1 data:Selection",
+          "    └─Selection 0.00 cop[tikv]  eq(test_partition.t2.b, 7), eq(test_partition.t2.id, 7), in(test_partition.t2.a, 6, 7, 8)",
```
Comment (AilinKid) [inline pkg/planner/core/casetest/partition/testdata/partition_pruner_out.json:477]: why the est changed is becoming more smaller? cause from the code logic is max(0, ?)

**Review by AilinKid — COMMENTED** [review]
Rest looks good to me

File: pkg/planner/core/casetest/planstats/testdata/plan_stats_suite_out.json [hunk 3]
```diff
@@ -130,16 +130,16 @@
...
+          "├─TableReader(Build) 1.00 root  data:TableRangeScan",
+          "│ └─TableRangeScan 1.00 cop[tikv] table:t2 range:(10,+inf], keep order:false, stats:partial[a:allEvicted]",
+          "└─IndexJoin(Probe) 0.33 root  inner join, inner:IndexLookUp, outer key:test.t.b, inner key:test.tp.c, equal cond:eq(test.t.b, test.tp.c)",
+          "  ├─TableReader(Build) 0.33 root  data:Selection",
+          "  │ └─Selection 0.33 cop[tikv]  gt(test.t.b, 10), not(isnull(test.t.b))",
+          "  │   └─TableRangeScan 1.00 cop[tikv] table:t range:[-inf,10), keep order:false, stats:partial[idx:allEvicted, a:allEvicted, b:allEvicted]",
+          "  └─IndexLookUp(Probe) 0.33 root partition:p0 ",
+          "    ├─Selection(Build) 0.33 cop[tikv]  gt(test.tp.c, 10), not(isnull(test.tp.c))",
+          "    │ └─IndexRangeScan 0.50 cop[tikv] table:tp, index:ic(c) range: decided by [eq(test.tp.c, test.t.b)], keep order:false, stats:partial[c:allEvicted]",
+          "    └─TableRowIDScan(Probe) 0.33 cop[tikv] table:tp keep order:false, stats:partial[c:allEvicted]"
```
Comment (AilinKid) [inline pkg/planner/core/casetest/planstats/testdata/plan_stats_suite_out.json:142]: same behavior

File: pkg/planner/core/casetest/partition/testdata/partition_pruner_out.json (hunk 2 above)
Comment (terry1purcell) [reply pkg/planner/core/casetest/partition/testdata/partition_pruner_out.json:477]: The estimates look to be the same, it is simply that the build & probe sides have changed - and not that the individual table estimates themselves have gotten smaller.

File: pkg/planner/core/casetest/planstats/testdata/plan_stats_suite_out.json (hunk 3 above)
Comment (terry1purcell) [reply pkg/planner/core/casetest/planstats/testdata/plan_stats_suite_out.json:142]: Each of the table row estimates are the same - it is simply that the order of operations has changed. Each individual table estimate did not increase or decrease.

**Review by AilinKid — APPROVED** [review]

**Pull request merged**